wallet-core/src/ffi.rs

```diff
@@ -619,3 +630,187 @@ pub unsafe fn moonlight_to_phoenix(
 
     ErrorCode::Ok
 }
+
+#[no_mangle]
+pub unsafe fn moonlight_stake(
+    seed: &Seed,
+    sender_index: u8,
+    stake_value: *const u64,
+    gas_limit: *const u64,
+    gas_price: *const u64,
+    nonce: *const u64,
+    chain_id: u8,
+    stake_nonce: *const u64,
+    tx_ptr: *mut *mut u8,
+    hash_ptr: &mut [u8; 64],
+) -> ErrorCode {
+    let transfer_value = 0;
+    let deposit = *stake_value;
+
+    let sender_sk = derive_bls_sk(&seed, sender_index);
+    let stake_sk = sender_sk.clone();
+
+    let stake = Stake::new(&stake_sk, *stake_value, *stake_nonce, chain_id);
+
+    let contract_call = ContractCall::new(STAKE_CONTRACT, "stake", &stake)
+        .or(Err(ErrorCode::ContractCallError))?;
+
+    let tx = crate::transaction::moonlight(
+        &sender_sk,
+        None,
+        transfer_value,
+        deposit,
+        *gas_limit,
+        *gas_price,
+        *nonce,
+        chain_id,
+        Some(contract_call),
+    )
+    .or(Err(ErrorCode::MoonlightTransactionError))?;
+
+    let bytes = tx.to_var_bytes();
+    let len = bytes.len().to_le_bytes();
+
+    let ptr = mem::malloc(4 + bytes.len() as u32);
+    let ptr = ptr as *mut u8;
+
+    *tx_ptr = ptr;
+
+    ptr::copy_nonoverlapping(len.as_ptr(), ptr, 4);
+    ptr::copy_nonoverlapping(bytes.as_ptr(), ptr.add(4), bytes.len());
+
+    let displayed = revert(&tx.hash());
+    let bytes = displayed.as_bytes();
+
+    ptr::copy_nonoverlapping(bytes.as_ptr(), hash_ptr.as_mut_ptr(), 64);
+
+    ErrorCode::Ok
```

Comment:
it seems that the part after the transaction creation can be moved to a different function to reduce code duplication

Reply:
Everything is duplicated at the moment on purpose; it will be refactored once the SDK works.

Reply:
As mentioned during the stand up I'm thinking how to reorganize the memory allocation between JS / workers / WebAssembly modules, that would impact this as well, so it will be probably be done in that refactoring.